src/goto-analyzer/show_on_source.cpp

```diff
@@ -0,0 +1,145 @@
+/*******************************************************************\
+
+Module: goto-analyzer
+
+Author: Daniel Kroening, kroening@kroening.com
+
+\*******************************************************************/
+
+#include "show_on_source.h"
+
+#include <util/file_util.h>
+#include <util/message.h>
+#include <util/unicode.h>
+
+#include <analyses/ai.h>
+
+#include <fstream>
+
+/// get the source files with non-top abstract states
+static std::set<std::string>
+get_source_files(const goto_modelt &goto_model, const ai_baset &ai)
+{
+  std::set<std::string> files;
+
+  for(const auto &f : goto_model.goto_functions.function_map)
+  {
+    forall_goto_program_instructions(i_it, f.second.body)
+    {
+      const auto abstract_state = ai.abstract_state_before(i_it);
+      if(abstract_state->is_top())
+        continue; // ignore
+
+      if(i_it->source_location.get_line().empty())
+        continue; // ignore
+
+      auto f = i_it->source_location.full_path();
+
+      if(!f.empty())
+        files.insert(f);
+    }
+  }
+
+  return files;
+}
+
+/// print a string with indent to match given sample line
+static void print_with_indent(
+  const std::string &src,
+  const std::string &indent_line,
+  std::ostream &out)
+{
+  const std::size_t p = indent_line.find_first_not_of(" \t");
+  const std::string indent =
+    p == std::string::npos ? std::string("") : std::string(indent_line, 0, p);
+  std::istringstream in(src);
+  std::string src_line;
+  while(std::getline(in, src_line))
+    out << indent << src_line << '\n';
+}
+
+/// output source code annotated with abstract states for given file
+void show_on_source(
+  const std::string &source_file,
+  const goto_modelt &goto_model,
+  const ai_baset &ai,
+  message_handlert &message_handler)
+{
+#ifdef _MSC_VER
+  std::ifstream in(widen(source_file));
+#else
+  std::ifstream in(source_file);
+#endif
+
+  messaget message(message_handler);
+
+  if(!in)
+  {
+    message.warning() << "Failed to open `" << source_file << "'"
+                      << messaget::eom;
+    return;
+  }
+
+  std::map<std::size_t, ai_baset::locationt> line_map;
+
+  // Collect line numbers with non-top abstract states.
+  // We pick the _first_ state for each line.
+  for(const auto &f : goto_model.goto_functions.function_map)
+  {
+    forall_goto_program_instructions(i_it, f.second.body)
+    {
+      auto abstract_state = ai.abstract_state_before(i_it);
+      if(abstract_state->is_top())
+        continue; // ignore
+
+      const auto line = i_it->source_location.get_line();
+
+      if(line.empty())
+        continue; // ignore
```

Comment:
There is a lot of overlap between this code and `get_source_files`. Could this be refactored?

Reply:
done